hazelcast/src/main/java/com/hazelcast/instance/NodeExtension.java

```diff
@@ -197,14 +197,38 @@
     void validateJoinRequest(JoinMessage joinMessage);
 
     /**
-     * Called when cluster state is changed
+     * Called before starting a cluster state change transaction.
+     *
+     * @param currState the state before the change
+     * @param requestedState the requested cluster state
+     * @param isTransient whether the change will be recorded in persistent storage, affecting the
+ *                    initial state after cluster restart. Transient changes happen during
+     */
+    void beforeClusterStateChange(ClusterState currState, ClusterState requestedState, boolean isTransient);
```

Comment:
I think you can explicitly mention that `beforeClusterStateChange(..)` and `afterClusterStateChange(..)` are called only on caller side which initiates the transaction. But `onClusterStateChange(..)` is called on all members.